Implement Integrity-Policy for scripts

yoavweiss opened the pull request:
#### 2745429641071660f77d69c64cd04727ee8e31c5
<pre>
Implement Integrity-Policy for scripts
<a href="https://bugs.webkit.org/show_bug.cgi?id=293433">https://bugs.webkit.org/show_bug.cgi?id=293433</a>

Reviewed by Sam Weinig.

This PR implements the Integrity-Policy header and enforces it on script destinations.

* LayoutTests/TestExpectations: Deleted expectations for Integrity-Policy tests.
* LayoutTests/imported/w3c/web-platform-tests/common/dispatcher/dispatcher.js: Imported the latest from WPT.
* LayoutTests/imported/w3c/web-platform-tests/common/dispatcher/remote-executor.html: Imported the latest from WPT.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/resources/ran.js: Imported the latest from WPT.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/remote-context-helper-tests/navigation-bfcache.window-expected.txt: Expectation change due to imports.
* LayoutTests/imported/w3c/web-platform-tests/mixed-content/csp.https.window-expected.txt: Progression due to imports.
* LayoutTests/imported/w3c/web-platform-tests/reporting/resources/report-helper.js:
(getReport): Ensured that blockedURL also counts towards getting a report.
* LayoutTests/imported/w3c/web-platform-tests/subresource-integrity/integrity-policy/parsing_type=enforce-expected.txt: Adjusted expectations.
* LayoutTests/imported/w3c/web-platform-tests/subresource-integrity/integrity-policy/parsing_type=report-expected.txt: Adjusted expectations.
* LayoutTests/imported/w3c/web-platform-tests/subresource-integrity/integrity-policy/script.https-expected.txt: Adjusted expectations.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/subresource-integrity/integrity-policy/script.https-expected.txt: Added failing expectation, due to https issues with reporting.
* LayoutTests/platform/mac-wk1/TestExpectations: Set expectations for reporting-based tests to fail on WK1.
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm: Include DocumentInlines.h from JS*Document cpp files.
* Source/WebCore/CMakeLists.txt: Add IntegrityPolicyViolationReportBody.idl
* Source/WebCore/DerivedSources-input.xcfilelist: Add IntegrityPolicyViolationReportBody.idl
* Source/WebCore/DerivedSources-output.xcfilelist: Add IntegrityPolicyViolationReportBody.idl
* Source/WebCore/DerivedSources.make: Add IntegrityPolicyViolationReportBody.idl
* Source/WebCore/Headers.cmake: Add IntegrityPolicyViolationReportBody.h
* Source/WebCore/Modules/reporting/IntegrityPolicyViolationReportBody.cpp: Added.
(WebCore::IntegrityPolicyViolationReportBody::IntegrityPolicyViolationReportBody):
(WebCore::IntegrityPolicyViolationReportBody::create):
(WebCore::IntegrityPolicyViolationReportBody::type const):
(WebCore::IntegrityPolicyViolationReportBody::documentURL const):
(WebCore::IntegrityPolicyViolationReportBody::blockedURL const):
(WebCore::IntegrityPolicyViolationReportBody::destination const):
(WebCore::IntegrityPolicyViolationReportBody::reportOnly const):
* Source/WebCore/Modules/reporting/IntegrityPolicyViolationReportBody.h: Added.
(isType):
* Source/WebCore/Modules/reporting/IntegrityPolicyViolationReportBody.idl:
* Source/WebCore/Modules/reporting/ReportingObserver.cpp:
(WebCore::isVisibleToReportingObservers): Add &quot;integrity-violation&quot; as a visible observer.
* Source/WebCore/Modules/reporting/ViolationReportType.h: Add IntegrityPolicy as a report type.
* Source/WebCore/Modules/url-pattern/URLPattern.cpp: Add required header include.
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.h: Add required header include.
* Source/WebCore/Modules/webdatabase/DatabaseManager.cpp: Add required header include.
* Source/WebCore/Modules/websockets/WorkerThreadableWebSocketChannel.h: Fix C++ safety issues.
* Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.cpp: Add required header include.
* Source/WebCore/Modules/webxr/WebXRBoundedReferenceSpace.cpp: Add required header include.
* Source/WebCore/Modules/webxr/WebXRSession.cpp: Add required header include.
* Source/WebCore/Sources.txt: Add IntegrityPolicyViolationReportBody.cpp (and its JS eq equivalent), and IntegrityPolicy.cpp
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSReportBodyCustom.cpp:
(WebCore::toJSNewlyCreated): Add IntegrityViolation handling.
* Source/WebCore/dom/SecurityContext.cpp: Define getters/setters.
(WebCore::SecurityContext::~SecurityContext):
(WebCore::SecurityContext::integrityPolicy const):
(WebCore::SecurityContext::setIntegrityPolicy):
(WebCore::SecurityContext::integrityPolicyReportOnly const):
(WebCore::SecurityContext::setIntegrityPolicyReportOnly):
* Source/WebCore/dom/SecurityContext.h: Add integrityPolicy and integrityPolicyReportOnly.
(WebCore::SecurityContext::integrityPolicy const):
(WebCore::SecurityContext::setIntegrityPolicy):
(WebCore::SecurityContext::integrityPolicyReportOnly const):
(WebCore::SecurityContext::setIntegrityPolicyReportOnly):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::responseReceived): Process the Integrity-Policy headers.
(WebCore::DocumentLoader::integrityPolicy): Getter impl.
(WebCore::DocumentLoader::integrityPolicyReportOnly): Getter impl.
* Source/WebCore/loader/DocumentLoader.h: Hold unique_ptr for integrityPolicy and integrityPolicyReportOnly and move them when requested.
(WebCore::DocumentLoader::integrityPolicy):
(WebCore::DocumentLoader::integrityPolicyReportOnly):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didBeginDocument): Set integrityPolicy and integrityPolicyReportOnly on Document.
* Source/WebCore/loader/IntegrityPolicy.cpp: Added.
(WebCore::reportViolation): Report an integrity violation.
(WebCore::processIntegrityPolicy): Parse and process an integrity policy header.
(WebCore::shouldRequestBeBlockedByIntegrityPolicy): Asses if a request should be blocked due to integrity policy.
* Source/WebCore/loader/IntegrityPolicy.h:
* Source/WebCore/loader/PingLoader.cpp:
(WebCore::PingLoader::sendViolationReport): Add IntegrityPolicy as a violation report type.
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::canRequest): call shouldRequestBeBlockedByIntegrityPolicy.
* Source/WebCore/platform/network/HTTPHeaderNames.in: Add Integrity-Policy and Integrity-Policy-Report-Only headers
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/295903@main">https://commits.webkit.org/295903@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e38cf0a29c75b176eb537ecc5f139ae739326e10

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106423 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26172 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16570 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111625 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/57019 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108462 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26839 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34675 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80839 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109427 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21290 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96043 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61167 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20764 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14141 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56460 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90620 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14176 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114485 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33561 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24752 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89915 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33925 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92271 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89617 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22881 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34503 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12321 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29170 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33486 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38898 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33232 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36585 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34830 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/Modules/reporting/IntegrityPolicyViolationReportBody.h:
```diff
@@ -0,0 +1,65 @@
+class IntegrityPolicyViolationReportBody final : public ReportBody {
+    WTF_MAKE_TZONE_OR_ISO_ALLOCATED(IntegrityPolicyViolationReportBody);
+public:
+    WEBCORE_EXPORT static Ref<IntegrityPolicyViolationReportBody> create(String&& documentURL, String&& blockedURL, String&& destination, bool reportOnly);
```
What is requiring these to be exported?

yoavweiss replied:
Nothing! removed

weinig commented on LayoutTests/imported/w3c/web-platform-tests/subresource-integrity/integrity-policy/script.https-expected.txt:
```diff
@@ -1,16 +1,14 @@
+PASS Ensure that a no-cors script gets blocked
+PASS Ensure that ReportingObserver gets called without endpoints
+PASS Ensure that a script with integrity runs
+FAIL Ensure that a script with signature integrity runs assert_equals: expected true but got false
```
Is signature integrity a separate specification? Is that planned?

yoavweiss replied:
It is a separate spec, and this test case should've been a tentative test. I can work on separating that upstream.

weinig commented on Source/WebCore/Modules/reporting/ViolationReportType.h:
```diff
@@ -36,6 +36,7 @@ enum class ViolationReportType : uint8_t {
     Deprecation, // https://wicg.github.io/deprecation-reporting/
     StandardReportingAPIViolation, // https://www.w3.org/TR/reporting/#try-delivery
     Test, // https://www.w3.org/TR/reporting-1/#generate-test-report-command
+    IntegrityPolicy,
```
Is there a URL that can go here to match the others above it?

yoavweiss replied:
Added

weinig commented on Source/WebCore/Modules/reporting/IntegrityPolicyViolationReportBody.cpp:
```diff
@@ -0,0 +1,77 @@
+
+const String& IntegrityPolicyViolationReportBody::documentURL() const
+{
+    return m_documentURL;
```
These trivial getters should go inline in the header.

yoavweiss replied:
Done!

weinig commented on Source/WebCore/loader/DocumentLoader.h:
```diff
@@ -43,6 +43,7 @@
 #include "FrameDestructionObserver.h"
 #include "FrameLoaderTypes.h"
 #include "HTTPSByDefaultMode.h"
+#include "IntegrityPolicy.h"
```
Can this #include be avoided with a forward declaration? Ideally, we wouldn't need to make IntegrityPolicy.h a private header.

yoavweiss replied:
Indeed!

weinig commented on Source/WebCore/loader/IntegrityPolicy.cpp:
```diff
@@ -0,0 +1,142 @@
+    if (!integrityPolicy && !integrityPolicyReportOnly)
+        return false;
+    // We don't currently support anything but script
+    if (options.destination != FetchOptionsDestination::Script)
```
I think there is currently support for integrity metadata on HTMLLinkElements for CSS style sheets, and support for integrity for CSS images is in the works. Will this check be problematic for either of those? (I realize the PR is about implementing this for scripts, so I am not asking for you to add more support, just checking that this won't break anything).

yoavweiss replied:
> and support for integrity for CSS images is in the works

I did not know that. Any pointers?

> Will this check be problematic for either of those? 

It won't, as non-script destinations would just pass this check and then fail at their actual integrity verifications (which happens at response time)

weinig replied:
> I did not know that. Any pointers?

The spec part is here -> https://www.w3.org/TR/css-values-5/#typedef-request-url-modifier-integrity-modifier

Initial implementation work was started in https://github.com/WebKit/WebKit/pull/43802 and https://github.com/WebKit/WebKit/pull/43910, but I haven't finished it off yet as it requires some work to make integrity work with with CachedImage.

yoavweiss replied:
Thanks!

One thing to watch out for here is that verifying integrity on images would prevent them from being progressively decoded, which may result in performance regressions.

weinig commented on Source/WebCore/Headers.cmake:
```diff
@@ -598,6 +598,7 @@ set(WebCore_PRIVATE_FRAMEWORK_HEADERS
     Modules/push-api/PushSubscriptionOwner.h
 
     Modules/reporting/DeprecationReportBody.h
+    Modules/reporting/IntegrityPolicyViolationReportBody.h
```
I don't think this header needs to be exposed. I don't it anywhere it is being included that would require this.

yoavweiss replied:
Removed

weinig commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -4821,6 +4821,8 @@
 		B8DBDB4E130B0F8A00F5CDB1 /* SpellingCorrectionCommand.h in Headers */ = {isa = PBXBuildFile; fileRef = B8DBDB4A130B0F8A00F5CDB1 /* SpellingCorrectionCommand.h */; };
 		B9CA3B402CAB1D3E00B2607C /* SpatialImageControls.h in Headers */ = {isa = PBXBuildFile; fileRef = B9CA3B3E2CAB1D3E00B2607C /* SpatialImageControls.h */; };
 		B9CA3B412CAB1D3E00B2607C /* SpatialImageControls.cpp in Sources */ = {isa = PBXBuildFile; fileRef = B9CA3B3F2CAB1D3E00B2607C /* SpatialImageControls.cpp */; };
+		BA5AA81B2DDF59210038722C /* IntegrityPolicyViolationReportBody.h in Headers */ = {isa = PBXBuildFile; fileRef = BA5AA8182DDF59100038722C /* IntegrityPolicyViolationReportBody.h */; settings = {ATTRIBUTES = (Private, ); }; };
```
I don't think this header needs to be exposed. I don't it anywhere it is being included that would require this. (Same comment as for Headers.cmake but for Xcode).

yoavweiss replied:
Removed

weinig commented on Source/WebCore/loader/IntegrityPolicy.cpp:
```diff
@@ -0,0 +1,142 @@
+        for (const auto& endpoint : endpoints) {
+            const URL& documentURL = document.url();
+
+            auto reportBody = IntegrityPolicyViolationReportBody::create(documentURL.string().isolatedCopy(), requestURL.string().isolatedCopy(), "script"_s, reportOnly);
```
The isolatedCopy() calls here are surprising. Why are they needed? Can you add a comment explaining that?

yoavweiss replied:
Removed

weinig commented on Source/WebCore/dom/SecurityContext.h:
```diff
@@ -83,6 +84,12 @@ class SecurityContext {
     virtual const CrossOriginOpenerPolicy& crossOriginOpenerPolicy() const { return m_crossOriginOpenerPolicy; }
     void setCrossOriginOpenerPolicy(const CrossOriginOpenerPolicy& crossOriginOpenerPolicy) { m_crossOriginOpenerPolicy = crossOriginOpenerPolicy; }
 
+    virtual const IntegrityPolicy* integrityPolicy() const { return m_integrityPolicy.get(); }
```
This doesn't seem to be overridden anywhere. Why does it need to be virtual?

yoavweiss replied:
It doesn't!

weinig commented on Source/WebCore/dom/SecurityContext.h:
```diff
@@ -83,6 +84,12 @@ class SecurityContext {
+    virtual const IntegrityPolicy* integrityPolicy() const { return m_integrityPolicy.get(); }
+    void setIntegrityPolicy(std::unique_ptr<IntegrityPolicy>&& policy) { m_integrityPolicy = WTFMove(policy); }
+
+    virtual const IntegrityPolicy* integrityPolicyReportOnly() const { return m_integrityPolicyReportOnly.get(); }
```
This doesn't seem to be overridden anywhere. Why does it need to be virtual?

yoavweiss replied:
It does not

weinig commented on Source/WebCore/loader/DocumentLoader.h:
```diff
@@ -97,6 +97,7 @@ struct CustomHeaderFields;
 class FormState;
 class FrameLoader;
 class IconLoader;
+struct IntegrityPolicy;
```
We usually sort the structs below the classes.

yoavweiss replied:
done

weinig reviewed (commented):
Looking good. Will do a final review when the bots are passing.

weinig commented on Source/WebCore/dom/SecurityContext.cpp:
```diff
@@ -37,7 +38,7 @@ namespace WebCore {
 SecurityContext::SecurityContext() = default;
 
-SecurityContext::~SecurityContext() = default;
+SecurityContext::~SecurityContext() { }
```
Why is this change needed?

yoavweiss replied:
I thought it's required for `std:unique_ptr` of forward references, but looks like it isn't. I'll revert.

yoavweiss committed:
Implement Integrity-Policy for scripts
https://bugs.webkit.org/show_bug.cgi?id=293433

Reviewed by Sam Weinig.

This PR implements the Integrity-Policy header and enforces it on script destinations.

* LayoutTests/TestExpectations: Deleted expectations for Integrity-Policy tests.
* LayoutTests/imported/w3c/web-platform-tests/common/dispatcher/dispatcher.js: Imported the latest from WPT.
* LayoutTests/imported/w3c/web-platform-tests/common/dispatcher/remote-executor.html: Imported the latest from WPT.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/resources/ran.js: Imported the latest from WPT.
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/remote-context-helper-tests/navigation-bfcache.window-expected.txt: Expectation change due to imports.
* LayoutTests/imported/w3c/web-platform-tests/mixed-content/csp.https.window-expected.txt: Progression due to imports.
* LayoutTests/imported/w3c/web-platform-tests/reporting/resources/report-helper.js:
(getReport): Ensured that blockedURL also counts towards getting a report.
* LayoutTests/imported/w3c/web-platform-tests/subresource-integrity/integrity-policy/parsing_type=enforce-expected.txt: Adjusted expectations.
* LayoutTests/imported/w3c/web-platform-tests/subresource-integrity/integrity-policy/parsing_type=report-expected.txt: Adjusted expectations.
* LayoutTests/imported/w3c/web-platform-tests/subresource-integrity/integrity-policy/script.https-expected.txt: Adjusted expectations.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/subresource-integrity/integrity-policy/script.https-expected.txt: Added failing expectation, due to https issues with reporting.
* LayoutTests/platform/mac-wk1/TestExpectations: Set expectations for reporting-based tests to fail on WK1.
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm: Include DocumentInlines.h from JS*Document cpp files.
* Source/WebCore/CMakeLists.txt: Add IntegrityPolicyViolationReportBody.idl
* Source/WebCore/DerivedSources-input.xcfilelist: Add IntegrityPolicyViolationReportBody.idl
* Source/WebCore/DerivedSources-output.xcfilelist: Add IntegrityPolicyViolationReportBody.idl
* Source/WebCore/DerivedSources.make: Add IntegrityPolicyViolationReportBody.idl
* Source/WebCore/Headers.cmake: Add IntegrityPolicyViolationReportBody.h
* Source/WebCore/Modules/reporting/IntegrityPolicyViolationReportBody.cpp: Added.
(WebCore::IntegrityPolicyViolationReportBody::IntegrityPolicyViolationReportBody):
(WebCore::IntegrityPolicyViolationReportBody::create):
(WebCore::IntegrityPolicyViolationReportBody::type const):
(WebCore::IntegrityPolicyViolationReportBody::documentURL const):
(WebCore::IntegrityPolicyViolationReportBody::blockedURL const):
(WebCore::IntegrityPolicyViolationReportBody::destination const):
(WebCore::IntegrityPolicyViolationReportBody::reportOnly const):
* Source/WebCore/Modules/reporting/IntegrityPolicyViolationReportBody.h: Added.
(isType):
* Source/WebCore/Modules/reporting/IntegrityPolicyViolationReportBody.idl:
* Source/WebCore/Modules/reporting/ReportingObserver.cpp:
(WebCore::isVisibleToReportingObservers): Add "integrity-violation" as a visible observer.
* Source/WebCore/Modules/reporting/ViolationReportType.h: Add IntegrityPolicy as a report type.
* Source/WebCore/Modules/url-pattern/URLPattern.cpp: Add required header include.
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.h: Add required header include.
* Source/WebCore/Modules/webdatabase/DatabaseManager.cpp: Add required header include.
* Source/WebCore/Modules/websockets/WorkerThreadableWebSocketChannel.h: Fix C++ safety issues.
* Source/WebCore/Modules/webtransport/WebTransportDatagramDuplexStream.cpp: Add required header include.
* Source/WebCore/Modules/webxr/WebXRBoundedReferenceSpace.cpp: Add required header include.
* Source/WebCore/Modules/webxr/WebXRSession.cpp: Add required header include.
* Source/WebCore/Sources.txt: Add IntegrityPolicyViolationReportBody.cpp (and its JS eq equivalent), and IntegrityPolicy.cpp
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSReportBodyCustom.cpp:
(WebCore::toJSNewlyCreated): Add IntegrityViolation handling.
* Source/WebCore/dom/SecurityContext.cpp: Define getters/setters.
(WebCore::SecurityContext::~SecurityContext):
(WebCore::SecurityContext::integrityPolicy const):
(WebCore::SecurityContext::setIntegrityPolicy):
(WebCore::SecurityContext::integrityPolicyReportOnly const):
(WebCore::SecurityContext::setIntegrityPolicyReportOnly):
* Source/WebCore/dom/SecurityContext.h: Add integrityPolicy and integrityPolicyReportOnly.
(WebCore::SecurityContext::integrityPolicy const):
(WebCore::SecurityContext::setIntegrityPolicy):
(WebCore::SecurityContext::integrityPolicyReportOnly const):
(WebCore::SecurityContext::setIntegrityPolicyReportOnly):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::responseReceived): Process the Integrity-Policy headers.
(WebCore::DocumentLoader::integrityPolicy): Getter impl.
(WebCore::DocumentLoader::integrityPolicyReportOnly): Getter impl.
* Source/WebCore/loader/DocumentLoader.h: Hold unique_ptr for integrityPolicy and integrityPolicyReportOnly and move them when requested.
(WebCore::DocumentLoader::integrityPolicy):
(WebCore::DocumentLoader::integrityPolicyReportOnly):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didBeginDocument): Set integrityPolicy and integrityPolicyReportOnly on Document.
* Source/WebCore/loader/IntegrityPolicy.cpp: Added.
(WebCore::reportViolation): Report an integrity violation.
(WebCore::processIntegrityPolicy): Parse and process an integrity policy header.
(WebCore::shouldRequestBeBlockedByIntegrityPolicy): Asses if a request should be blocked due to integrity policy.
* Source/WebCore/loader/IntegrityPolicy.h:
* Source/WebCore/loader/PingLoader.cpp:
(WebCore::PingLoader::sendViolationReport): Add IntegrityPolicy as a violation report type.
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::canRequest): call shouldRequestBeBlockedByIntegrityPolicy.
* Source/WebCore/platform/network/HTTPHeaderNames.in: Add Integrity-Policy and Integrity-Policy-Report-Only headers
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/295903@main
